src/dbnode/encoding/tile/types.go

```diff
@@ -0,0 +1,105 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tile
+
+import (
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/dbnode/persist/fs"
+	"github.com/m3db/m3/src/dbnode/ts"
+	"github.com/m3db/m3/src/x/ident"
+	xtime "github.com/m3db/m3/src/x/time"
+)
+
+// SeriesFrameUnits describes units in this series frame.
+type SeriesFrameUnits interface {
+	// SingleValue returns the last unit seen, and a boolean indicating if
+	// that unit is constant for all datapoints accross the series frame.
+	SingleValue() (xtime.Unit, bool)
+
+	// Values returns all values seen.
+	// NB: if this is called on a recorder with a constant unit, this will
+	// generate a full slice filled with that unit, corresponding to the
+	// number of units added to the recorder. It is recommended to call
+	// SingleValue first to avoid unnecessary allocs if SingleValue is true.
+	Values() []xtime.Unit
+}
+
+// SeriesFrameAnnotations describes annotations in this series frame.
+type SeriesFrameAnnotations interface {
+	// SingleValue returns the last annotation seen, and a boolean indicating if
+	// that annotation is constant for all datapoints accross the series frame.
+	SingleValue() (ts.Annotation, bool)
+
+	// Values returns all values seen.
+	// NB: if this is called on a recorder with a constant annotation, this will
+	// generate a full slice filled with that annotation, corresponding to the
+	// number of annotations added to the recorder. It is recommended to call
+	// SingleValue first to avoid unnecessary allocs if SingleValue is true.
+	Values() []ts.Annotation
+}
+
+// SeriesFrameIterator is a frame-wise iterator across a series block.
+type SeriesFrameIterator interface {
+	// Err returns any errors encountered.
+	Err() error
+	// Next moves to the next element.
+	Next() bool
+	// Close closes the iterator.
+	Close() error
+	// Current returns the current series block frame.
+	Current() SeriesBlockFrame
+	// Reset resets the series frame iterator.
+	Reset(
+		start xtime.UnixNano,
+		step xtime.UnixNano,
+		it fs.CrossBlockIterator,
+	) error
+}
+
+// SeriesBlockIterator provides concurrent iteration across multiple series
+// in a frame-wise fashion, exposing results as arrow slices.
+type SeriesBlockIterator interface {
+	// Err returns any errors encountered.
+	Err() error
+	// Next moves to the next element.
+	Next() bool
+	// Close closes the iterator.
+	Close() error
+	// Current returns the next set of series frame iterators, with relevant
+	// tags and IDs.
+	Current() (SeriesFrameIterator, ident.ID, []byte)
+}
+
+type recorder interface {
```

Comment:
Can probably drop it as we have only one implementation?

Reply:
Yeah refactored this 👍